quinn-proto/src/frame.rs

```diff
@@ -544,21 +546,23 @@ impl From<UnexpectedEnd> for IterErr {
 }
 
 impl Iter {
-    pub fn new(payload: Bytes) -> Self {
+    pub fn new(payload: BytesMut) -> Self {
         Iter {
             bytes: io::Cursor::new(payload),
             last_ty: None,
         }
     }
 
-    fn take_len(&mut self) -> Result<Bytes, UnexpectedEnd> {
+    fn take_len(&mut self) -> Result<BytesMut, UnexpectedEnd> {
         let len = self.bytes.get_var()?;
         if len > self.bytes.remaining() as u64 {
             return Err(UnexpectedEnd);
         }
         let start = self.bytes.position() as usize;
         self.bytes.advance(len as usize);
-        Ok(self.bytes.get_ref().slice(start..(start + len as usize)))
+        Ok(BytesMut::from(
```

Comment:
I think this now creates a deep copy of all incoming data, where there hasn't been one before.

Since includes datagram frames, it should effectively negate the whole purpose of this PR?

My guess is that it didn't show up more negative in the benchmark since that one just uses a single stream and has always sufficient data to send, to the application runs into the `take_remaining` case which stays cheap.  But if people try to send smaller datagrams or use small streams it would it get more expensive.

Reply:
> Since includes datagram frames, it should effectively negate the whole purpose of this PR?

True. I wasn't aware ``slice`` was a cheap operation; I somehow thought it was a copy.

Going through this all, it seems this PR won't work without significantly changing the buffering mechanism. Have any ideas?

Reply:
It seems like one thing we could do is to buffer `BytesMut`s throughout the (quinn-proto) stack. That way we can split them wherever necessary, I think, without copying. Would that make sense?